api/api.go

```diff
@@ -94,3 +133,50 @@ func (api *API) Update(cfg *config.Config, resolveTimeout time.Duration) error {
 
 	return api.v2.Update(cfg, resolveTimeout)
 }
+
+// makeLimiter returns an HTTP middleware that sets a timeout for HTTP requests
+// and also limits the number of concurrently processed GET requests to the
+// given number.
+//
+// If timeout is < 1, no timeout is enforced.
+//
+// If concurrency is < 1, GOMAXPROCS is used as the concurrency limit but at least 8.
+//
+// The returned middleware serves http.StatusServiceUnavailable (503) for requests that
+// would exceed the number.
+func makeLimiter(timeout time.Duration, concurrency int) func(http.Handler) http.Handler {
+	if concurrency < 1 {
+		concurrency = runtime.GOMAXPROCS(0)
+		if concurrency < 8 {
+			concurrency = 8
+		}
+	}
+	inFlightSem := make(chan struct{}, concurrency)
+
+	return func(h http.Handler) http.Handler {
+		concLimiter := http.HandlerFunc(func(rsp http.ResponseWriter, req *http.Request) {
+			if req.Method == http.MethodGet { // Only limit concurrency of GETs.
+				select {
+				case inFlightSem <- struct{}{}: // All good, carry on.
+					requestsInFlight.Inc()
+					defer func() {
+						<-inFlightSem
+						requestsInFlight.Dec()
+					}()
+				default:
+					http.Error(rsp, fmt.Sprintf(
+						"Limit of concurrent GET requests reached (%d), try again later.\n", concurrency,
+					), http.StatusServiceUnavailable)
+					return
+				}
+			}
+			h.ServeHTTP(rsp, req)
+		})
+		if timeout <= 0 {
+			return concLimiter
+		}
+		return http.TimeoutHandler(concLimiter, timeout, fmt.Sprintf(
+			"Exceeded configured timeout of %v.\n", timeout,
```

Comment:
```suggestion
			"exceeded configured timeout of %v.\n", timeout,
```

Reply:
That's actually not a Go error message (which, by convention, is lower case) but the content of the HTTP response. As such, it is shown in isolation and should follow normal English spelling rules. (The default error message of the TimeoutHandler is also capitalized. It is even using HTML, but I don't think we have to go that far.)